File: Blind75-Practice/easy/05-sliding-window/Python/contains-duplicate-ii.py
"""
LeetCode 219: Contains Duplicate II
Difficulty: Easy
Pattern: Sliding Window

Given an integer array nums and an integer k, return true if there are two distinct indices
i and j such that nums[i] == nums[j] and abs(i - j) <= k.

Approach: Hash map from value to last index; slide by removing stale entries or checking distance.

Time Complexity: O(n)
Space Complexity: O(min(n, k)) — at most k+1 distinct indices matter in the window
"""

from typing import Dict, List


class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        last: Dict[int, int] = {}
        for i, x in enumerate(nums):
            if x in last and i - last[x] <= k:
                return True
            last[x] = i
        return False


if __name__ == "__main__":
    s = Solution()
    assert s.containsNearbyDuplicate([1, 2, 3, 1], 3) is True
    assert s.containsNearbyDuplicate([1, 0, 1, 1], 1) is True
    assert s.containsNearbyDuplicate([1, 2, 3, 1, 2, 3], 2) is False
    print("All test cases passed!")

File: Blind75-Practice/easy/05-sliding-window/Java/ContainsDuplicateII.java
/*
219. Contains Duplicate II
Difficulty: Easy
Pattern: Sliding Window

Return true if same value appears twice with index distance at most k.

Approach: HashMap from value to last index; update while scanning.

Time: O(n), Space: O(min(n, k))
*/

public class ContainsDuplicateII {

    // [1 4 2 3 4 4 1 2 3 4] k = 3
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        //

        // Set<Integer> seen = new HashSet<>();
        // for (int i = 0; i < nums.length; i++) {
        //     if (!seen.add(nums[i])) return true;
        //     if (seen.size() > k) seen.remove(nums[i - k]);
        // }
        // return false;

    }

    public static void main(String[] args) {
        ContainsDuplicateII s = new ContainsDuplicateII();
        assert s.containsNearbyDuplicate(new int[] {1, 2, 3, 1}, 3);
        assert s.containsNearbyDuplicate(new int[] {1, 0, 1, 1}, 1);
        assert !s.containsNearbyDuplicate(new int[] {1, 2, 3, 1, 2, 3}, 2);
        System.out.println("All test cases passed!");
    }
}
